src/NuGet.Core/NuGet.ProjectModel/HashWriter.cs

```diff
@@ -0,0 +1,223 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using NuGet.Common;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// Generates a hash from an object graph.
+    /// 
+    /// This is non-private only to facilitate unit testing.
+    /// </summary>
+    public sealed class HashWriter : IObjectWriter, IDisposable
+    {
+        private const int _defaultBufferSize = 4096;
+
+        // These delimiters are important in preserving the structure of an object map.
+        // Although the values were chosen out of familiarly with JSON, other values could be used.
+        private static readonly byte[] _null = new[] { (byte)0 };
```

Comment:
This is not serialization.  Serialization is a reversible operation (via deserialization).  This operation is one way.

Nevertheless, your underlying concern is that when we later check if the dependency graph has changed via hash comparison this approach may lead to false negatives.

I updated the PR to use `JsonTextWriter` to incrementally produce JSON for hashing.  On a large project there is an average memory increase of ~100 KB (~30%).  Not as good as before, but still a significant improvement over the current solution.